frontends/p4/structInitializers.h

```diff
@@ -33,17 +34,36 @@ class CreateStructInitializers : public Transform {
         CHECK_NULL(refMap); CHECK_NULL(typeMap);
     }
 
-    const IR::Node* postorder(IR::AssignmentStatement* statement) override;
     const IR::Node* postorder(IR::MethodCallExpression* expression) override;
     const IR::Node* postorder(IR::Operation_Relation* expression) override;
 };
 
+class CreateStructAssignInitializers : public Transform {
+    ReferenceMap* refMap;
+    TypeMap* typeMap;
+ public:
+    CreateStructAssignInitializers(ReferenceMap* refMap, TypeMap* typeMap):
+            refMap(refMap), typeMap(typeMap) {
+        setName("CreateStructAssignInitializers");
+        CHECK_NULL(refMap); CHECK_NULL(typeMap);
+    }
+
+    const IR::Node* postorder(IR::AssignmentStatement* statement) override;
+};
+
 class StructInitializers : public PassManager {
  public:
-    StructInitializers(ReferenceMap* refMap, TypeMap* typeMap) {
+    StructInitializers(ReferenceMap* refMap, TypeMap* typeMap, bool doProcessAssignments) {
         setName("StructInitializers");
         passes.push_back(new TypeChecking(refMap, typeMap));
-        passes.push_back(new CreateStructInitializers(refMap, typeMap));
+        if (doProcessAssignments == false) {
+            passes.push_back(new CreateStructInitializers(refMap, typeMap));
+        } else {
+            passes.push_back(new CreateStructAssignInitializers(refMap, typeMap));
+            passes.push_back(new ResolveReferences(refMap));
+            // may insert new constants
+            passes.push_back(new TypeInference(refMap, typeMap, false));
```

Comment:
But aren't all inserted constants having the correct type already?
I suspect this may not be necessary. Can you try without it as well? If I am right, the ResolveReferences is not necessary either.

Reply:
When I try to compile programs in which mix of explicit
values and default values are used (e.g ` h={3, true, ...}`), without those two additional passes I get compiler bug:
p4c/frontends/p4/typeChecking/typeChecker.cpp:144
Compiler Bug: At this point in the compilation typechecking should not infer new types anymore, but it did.

I suppose that is because constants used for explicit initialization of fields don't have the correct type already. So ResolveReferences and TypeInference are needed, am I right?

Reply:
Then you should add a comment that these passes are useful not for the newly introduced initializers, but for the existing ones. I was assuming these had already been processed, but I guess the first instance of StructIntializers with a 'false' flag no longer does that.

Reply:
They are useful only if a mix of explicit and default values are used, because type unification of struct fields and explicit values in list expression listed before `...` is postponed after StructInitializers pass (with true flag). That is why those constants do not have correct type already. 
Should I try to make some changes in typeUnification.cpp and make these passes unnecessary again, or should I leave it like this (with added comment, as you said)?

Reply:
The important thing is for passes in the front-end to use the correct assumptions; none of the passes before the second invocation of StructInitializers should assume that constants have been cast implicitly, for example. In general we try to minimize the number of assumptions that pass writers have to understand, so if there could be only one version of StructInitializers it would be better - but if the trade-off is increased complexity then it may not be worth it.